### PR TITLE
unused_code: handle git grep no-matches and untracked

### PR DESCRIPTION
**Pull request by myakove**
detect pytest fixtures incl. param usage; fail CI on unexpected git errors; aggregate all unused functions per file; silence coverage show_contexts warning; fix jira_utils config precedence for tests

fix [186](https://github.com/RedHatQE/python-utility-scripts/issues/186)

## Summary by CodeRabbit

- New Features
  - Unused-code analyzer: analyze single files or directories, improved detection for pytest fixtures and usage patterns, auto-detects grep capabilities, deterministic sorted output, enhanced CLI options.

- Documentation
  - Expanded README with usage examples, CLI options, config examples, and skip annotations.

- Tests
  - Large suite of new tests and manifests covering fixtures, grep behavior, path handling, CLI flows, and edge cases.

- Chores
  - .gitignore appended with AI-related patterns; coverage HTML context display disabled.

- Style
  - Minor whitespace cleanup.

### CODE REVIEW

**Comment by redhat-qe-bot1** [comment]
Report bugs in [Issues](https://github.com/myakove/github-webhook-server/issues)

## Welcome! 🎉

This pull request will be automatically processed with the following features:

### 🔄 Automatic Actions
* **Reviewer Assignment**: Reviewers are automatically assigned based on the OWNERS file in the repository root
* **Size Labeling**: PR size labels (XS, S, M, L, XL, XXL) are automatically applied based on changes
* **Issue Creation**: A tracking issue is created for this PR and will be closed when the PR is merged or closed
* **Pre-commit Checks**: [pre-commit](https://pre-commit.ci/) runs automatically if `.pre-commit-config.yaml` exists
* **Branch Labeling**: Branch-specific labels are applied to track the target branch
* **Auto-verification**: Auto-verified users have their PRs automatically marked as verified

### 📋 Available Commands

#### PR Status Management
* `/wip` - Mark PR as work in progress (adds WIP: prefix to title)
* `/wip cancel` - Remove work in progress status
* `/hold` - Block PR merging (approvers only)
* `/hold cancel` - Unblock PR merging
* `/verified` - Mark PR as verified
* `/verified cancel` - Remove verification status

#### Review & Approval
* `/lgtm` - Approve changes (looks good to me)
* `/approve` - Approve PR (approvers only)
* `/automerge` - Enable automatic merging when all requirements are met (maintainers and approvers only)
* `/assign-reviewers` - Assign reviewers based on OWNERS file
* `/assign-reviewer @username` - Assign specific reviewer
* `/check-can-merge` - Check if PR meets merge requirements

#### Testing & Validation
 * `/retest tox` - Run Python test suite with tox
 * `/retest python-module-install` - Test Python package installation
 * `/retest all` - Run all available tests


#### Container Operations
* `/build-and-push-container` - Build and push container image (tagged with PR number)
  * Supports additional build arguments: `/build-and-push-container --build-arg KEY=value`

#### Cherry-pick Operations
* `/cherry-pick <branch>` - Schedule cherry-pick to target branch when PR is merged
  * Multiple branches: `/cherry-pick branch1 branch2 branch3`

#### Label Management
* `/<label-name>` - Add a label to the PR
* `/<label-name> cancel` - Remove a label from the PR

### ✅ Merge Requirements

This PR will be automatically approved when the following conditions are met:

1. **Approval**: `/approve` from at least one approver
2. **LGTM Count**: Minimum 0 `/lgtm` from reviewers
3. **Status Checks**: All required status checks must pass
4. **No Blockers**: No WIP, hold, or conflict labels
5. **Verified**: PR must be marked as verified (if verification is enabled)

### 📊 Review Process

<details>
<summary><strong>Approvers and Reviewers</strong></summary>


**Approvers:**
 * dbasunag
 * myakove
 * rnetser


**Reviewers:**
 * dbasunag
 * myakove
 * rnetser


</details>

<details>
<summary><strong>Available Labels</strong></summary>

 * hold
 * verified
 * wip
 * lgtm
 * approve
 * automerge

</details>

### 💡 Tips

* **WIP Status**: Use `/wip` when your PR is not ready for review
* **Verification**: The verified label is automatically removed on each new commit
* **Cherry-picking**: Cherry-pick labels are processed when the PR is merged
* **Container Builds**: Container images are automatically tagged with the PR number
* **Permission Levels**: Some commands require approver permissions
* **Auto-verified Users**: Certain users have automatic verification and merge privileges

For more information, please refer to the project documentation or contact the maintainers.

File: apps/unused_code/unused_code.py [hunk 1]
```diff
@@ -31,6 +69,151 @@ def is_fixture_autouse(func: ast.FunctionDef) -> bool:
...
+        return True
+
+    # Pattern 2: Lines that look like type annotations in docstrings
+    # Must have descriptive text after the colon, not just code
+    type_annotation_pattern = rf"\b{re.escape(function_name)}\s*\([^)]*\)\s*:\s+[A-Z][a-z]"
+    if re.search(type_annotation_pattern, stripped_line):
+        return True
+
+    # Pattern 3: Lines that contain common documentation keywords near the function name
+    doc_keywords = ["Args:", "Arguments:", "Parameters:", "Returns:", "Return:", "Raises:", "Note:", "Example:"]
```
Comment (dbasunag) [inline apps/unused_code/unused_code.py:167]: These are typically not in the same line as the function names. Doc patterns typically includes one of these string and next line may have an arg name. So I am not sure if we will get to these lines.

File: apps/unused_code/unused_code.py [hunk 2]
```diff
@@ -31,6 +69,151 @@ def is_fixture_autouse(func: ast.FunctionDef) -> bool:
...
+            # If we find doc keywords and function name with parens, likely documentation
+            if f"{function_name}(" in stripped_line:
+                return True
+            # Also catch cases where the keyword line itself contains the function name
+            if keyword.rstrip(":").lower() == function_name.lower():
+                return True
+
+    # Pattern 4: Check for common docstring patterns
+    # Lines that start with common documentation patterns
+    doc_starters = ['"""', "'''", "# ", "## ", "### ", "*", "-", "•"]
```
Comment (dbasunag) [inline apps/unused_code/unused_code.py:179]: If I am calling a member function, that would include . e.g. <resource>.wait_for_status(). I may have a resource class in my repo.

File: apps/unused_code/unused_code.py [hunk 3]
```diff
@@ -31,6 +69,151 @@ def is_fixture_autouse(func: ast.FunctionDef) -> bool:
...
+        return rf"\b{function_name}\s*[(]"
+    # -G basic regex: use word-start token \< and literal '('
+    return rf"\<{function_name}[[:space:]]*[(]"
+
+
+def _build_fixture_param_pattern(function_name: str) -> str:
+    r"""Build a portable regex to find a parameter named `function_name` in a def signature."""
+    flag = _detect_supported_grep_flag()
+    if flag == "-P":
+        return rf"def\s+\w+\s*[(][^)]*\b{function_name}\b"
```
Comment (dbasunag) [inline apps/unused_code/unused_code.py:120]: This would find a fixture if it appears in the first line of the test or fixture definition. How about those who appears in a different line? e.g.
```
 @pytest.mark.usefixtures("sa_namespace", "service_account", "mr_access_role", "mr_access_role_binding")
    def test_service_account_access_granted(
        self: Self,
        sa_token: str,
        model_registry_instance_rest_endpoint: list[str],
    ):

File: apps/unused_code/unused_code.py [hunk 4]
```diff
@@ -75,25 +260,59 @@ def process_file(py_file: str, func_ignore_prefix: list[str], file_ignore_list:
...
+            if _is_documentation_pattern(_line, func.name):
+                continue
+
             if func.name in _line:
                 used = True
                 break
 
+        # If not found and it's a pytest fixture, also search for parameter usage in function definitions
+        if not used and is_pytest_fixture(func=func):
+            param_pattern = _build_fixture_param_pattern(function_name=func.name)
```
Comment (dbasunag) [inline apps/unused_code/unused_code.py:293]: We should also look for fixtures in usefixtures()

**Comment by myakove** [comment]
@dbasunag Moved all function to test into unused_code_file_for_test.py for better visibility, check it and let me know if your concerns are still valid.

**Comment by myakove** [comment]
@dbasunag Please check again with the latest code

**Review by dbasunag — DISMISSED** [review]
Thanks @myakove !
